keps/sig-storage/1979-object-storage-support/README.md

```diff
@@ -0,0 +1,730 @@
+
+# Proposal
+
+## APIs
+
+### Storage APIs
+
+#### BucketRequest
+
+A user facing, namespaced custom resource requesting provisioning of a new bucket, or requesting access to an existing bucket. A `BucketRequest` (BR) lives in the app's namespace.  In addition to a `BucketRequest`, a [BucketAccessRequest](#bucketaccessrequest) is necessary in order to grant credentials to access the bucket. BRs are required for both greenfield and brownfield uses.
+
+There is a 1:1 mapping of a `BucketRequest` and the cluster scoped `Bucket` instance, meaning there is a single `Bucket` instance for every BR.
+
+```yaml
+apiVersion: cosi.io/v1alpha1
+kind: BucketRequest
+metadata:
+  name:
+  namespace:
+  labels:
+    - cosi.io/provisioner: [1]
+  finalizers:
+    - cosi.io/finalizer [2]
+spec:
+  protocol:
+    name: [3]
+    version: [4]
+  bucketPrefix: [5]
+  bucketClassName: [6]
+  bucketInstanceName: [7]
+status:
+  bucketAvailable: [8]
+```
+
+1. `labels`: added by COSI.  Key’s value should be the provisioner name. Characters that do not adhere to [Kubernetes label conventions](https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/#syntax-and-character-set) will be converted to ‘-’.
+1. `finalizers`: added by COSI to defer `BucketRequest` deletion until backend deletion succeeds.
+1. `protocol.name`: (required) specifies the desired protocol.  One of {“s3”, “gs”, or “azureBlob”}.
+1. `protocol.version`: (optional) specifies the desired version of the `protocol`. For "s3", a value of "v2" or "v4" could be used. 
+1. `bucketPrefix`: (optional) prefix prepended to a generated new bucket name, eg. “yosemite-photos-". If `bucketInstanceName` is supplied then `bucketPrefix` is ignored because the request is for access to an existing bucket. If `bucketPrexix` is specified then it is added to the `Bucket` instance as a label.
+1. `bucketClassName`: (optional for greenfield, omitted for brownfield) name of the `BucketClass` used to provision this request. If omitted for a greenfield bucket request, a default bucket class matching the protocol, if available, is used. If the greenfield bucket class is missing or does not support the requested protocol, an error is logged and the request is retried (with exponential backoff). A `BucketClass` is necessary for greenfield requests since BCs support a list of allowed namespaces. A `BucketClass` is not needed for brownfield requests since the `Bucket` instance, created by the admin, also contains `allowedNamespaces`.
+1. `bucketInstanceName`: (optional) name of the cluster-wide `Bucket` instance. If blank, then COSI assumes this is a greenfield request and will fill in the name during the binding step. If defined by the user, then this names the `Bucket` instance created by the admin. There is no automation available to make this name known to the user.
```

Comment:
If both `bucketPrefix` and `BucketInstanceName` in BucketRequest CR are empty, whether COSI will create a bucket with a random name?

Reply:
@thotz This is already explained in the [KEP](#bucket) . The BucketInstanceName is filled in by COSI, and the bucketPrefix is optional. The generated name follows the format br-${uuid}

Reply:
FIX: delete "There is no automation available to make this name known to the user."